File: keyple-plugins/android-omapi/src/main/java/org/eclipse/keyple/plugin/android/omapi/AndroidOmapiReader.java
/*
 * Copyright (c) 2018 Calypso Networks Association https://www.calypsonet-asso.org/
 *
 * All rights reserved. This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License version 2.0 which accompanies this distribution, and is
 * available at https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html
 */

package org.eclipse.keyple.plugin.android.omapi;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import org.eclipse.keyple.seproxy.SeProtocol;
import org.eclipse.keyple.seproxy.event.AbstractStaticReader;
import org.eclipse.keyple.seproxy.exception.ChannelStateReaderException;
import org.eclipse.keyple.seproxy.exception.IOReaderException;
import org.eclipse.keyple.seproxy.exception.NoStackTraceThrowable;
import org.eclipse.keyple.seproxy.protocol.ContactsProtocols;
import org.eclipse.keyple.util.ByteBufferUtils;
import org.simalliance.openmobileapi.Channel;
import org.simalliance.openmobileapi.Reader;
import org.simalliance.openmobileapi.Session;
import android.util.Log;

/**
 * Communicates with Android readers throught the Open Mobile API see {@link Reader} Instances of
 * this class represent SE readers supported by this device. These readers can be physical devices
 * or virtual devices. They can be removable or not. They can contain one SE that can or cannot be
 * removed.
 */
public class AndroidOmapiReader extends AbstractStaticReader {


    private static final String TAG = AndroidOmapiReader.class.getSimpleName();

    private Reader omapiReader;
    private Channel openChannel = null;
    private ByteBuffer openApplication = null;
    private Map<String, String> parameters = new HashMap<String, String>();

    protected AndroidOmapiReader(Reader omapiReader, String name) {
        super(name);
        this.omapiReader = omapiReader;
    }

    // @Override
    /**
     * Return the name of this reader. If this reader is a SIM reader, then its name must be
     * "SIM[Slot]" If the reader is a SD or micro SD reader, then its name must be “SD[slot]” If the
     * reader is an embedded SE reader, then its name must be “eSE[slot]” see
     * {@link Reader#getName()}
     *
     */
    /*
     * public String getName() { return omapiReader.getName(); }
     */

    @Override
    public Map<String, String> getParameters() {
        Log.w(TAG, "No parameters are supported by AndroidOmapiReader");
        return parameters;
    }

    @Override
    public void setParameter(String key, String value) throws IOException {
        Log.w(TAG, "No parameters are supported by AndroidOmapiReader");
        parameters.put(key, value);
    }

    /**
     * Check if a SE is present in this reader. see {@link Reader#isSecureElementPresent()}
     * 
     * @return True if the SE is present, false otherwise
     * @throws IOReaderException
     */
    @Override
    public boolean isSePresent() throws NoStackTraceThrowable {
        return omapiReader.isSecureElementPresent();
    }

    /**
     * Open a Channel to the application AID if not open yet. see {@link Reader#openSession()} see
     * {@link Session#openLogicalChannel(byte[])}
     * 
     * @param aid : aid from where to open the channel
     * @return Array : index[0] : ATR and index[1] :FCI
     * @throws IOReaderException
     */
    @Override
    protected ByteBuffer[] openLogicalChannelAndSelect(ByteBuffer aid,
            Set<Short> successfulSelectionStatusCodes) throws IOReaderException {
        ByteBuffer[] atrAndFci = new ByteBuffer[2];

        try {

            if (openChannel != null && !openChannel.isClosed() && openApplication != null
                    && openApplication.equals(aid)) {
                Log.i(TAG, "Channel is already open to aid : " + ByteBufferUtils.toHex(aid));

                atrAndFci[0] = ByteBuffer.wrap(openChannel.getSession().getATR());
                atrAndFci[1] = ByteBuffer.wrap(openChannel.getSelectResponse());


            } else {

                Log.i(TAG, "Opening channel to aid : " + ByteBufferUtils.toHex(aid));

                // open physical channel
                Session session = omapiReader.openSession();

                // get ATR from session
                Log.i(TAG, "Retrieveing ATR from session...");
                atrAndFci[0] =
                        ByteBuffer.wrap(session.getATR() != null ? session.getATR() : new byte[0]);

                Log.i(TAG, "Create logical openChannel within the session...");
                openChannel = session.openLogicalChannel(ByteBufferUtils.toBytes(aid));

                // get FCI
                atrAndFci[1] = ByteBuffer.wrap(openChannel.getSelectResponse());

            }
        } catch (IOException e) {
            throw new IOReaderException(e.getMessage(), e.getCause());
        } catch (SecurityException e) {
            throw new IOReaderException(e.getMessage(), e.getCause());
        }

        return atrAndFci;
    }

    /**
     * Close session see {@link Session#close()}
     * 
     * @throws IOReaderException
     */
    @Override
    protected void closePhysicalChannel() throws IOReaderException {
        openChannel.getSession().close();
        openChannel = null;
        openApplication = null;
    }

    /**
     * Transmit an APDU command (as per ISO/IEC 7816) to the SE see {@link Channel#transmit(byte[])}
     * 
     * @param apduIn byte buffer containing the ingoing data
     * @return
     * @throws ChannelStateReaderException
     */
    @Override
    protected ByteBuffer transmitApdu(ByteBuffer apduIn) throws ChannelStateReaderException {
        // Initialization
        Log.d(TAG, "Data Length to be sent to tag : " + apduIn.limit());
        Log.d(TAG, "Data in : " + ByteBufferUtils.toHex(apduIn));
        byte[] data = ByteBufferUtils.toBytes(apduIn);
        byte[] dataOut = new byte[0];
        try {
            dataOut = openChannel.transmit(data);
        } catch (IOException e) {
            e.printStackTrace();
            throw new ChannelStateReaderException(e);
        }
        ByteBuffer out = ByteBuffer.wrap(dataOut);
        Log.d(TAG, "Data out : " + ByteBufferUtils.toHex(out));
        return out;
    }

    /**
     * The only protocol Fla
     * 
     * @param protocolFlag
     * @return true
     * @throws IOReaderException
     */
    @Override
    protected boolean protocolFlagMatches(SeProtocol protocolFlag) throws IOReaderException {
        return protocolFlag.equals(ContactsProtocols.PROTOCOL_ISO7816_3);
    }
}

File: android/keyple-plugin/android-omapi/src/main/kotlin/org/eclipse/keyple/plugin/android/omapi/simalliance/AndroidOmapiReader.kt
/********************************************************************************
 * Copyright (c) 2020 Calypso Networks Association https://www.calypsonet-asso.org/
 *
 * See the NOTICE file(s) distributed with this work for additional information regarding copyright
 * ownership.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse
 * Public License 2.0 which is available at http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
package org.eclipse.keyple.plugin.android.omapi.simalliance

import java.io.IOException
import org.eclipse.keyple.core.service.exception.KeypleReaderIOException
import org.eclipse.keyple.core.util.ByteArrayUtil
import org.eclipse.keyple.plugin.android.omapi.AbstractAndroidOmapiReader
import org.eclipse.keyple.plugin.android.omapi.AndroidOmapiSupportedProtocols
import org.simalliance.openmobileapi.Channel
import org.simalliance.openmobileapi.Reader
import org.simalliance.openmobileapi.Session
import timber.log.Timber

/**
 * Implementation of the [AbstractAndroidOmapiReader] based on the [AbstractLocalReader]
 * with org.simalliance.omapi
 */
internal class AndroidOmapiReader(private val nativeReader: Reader, pluginName: String, readerName: String) :
    AbstractAndroidOmapiReader(pluginName, readerName) {

    companion object {
        private const val P2_SUPPORTED_MIN_VERSION = 3
    }

    private var session: Session? = null
    private var openChannel: Channel? = null
    private val omapiVersion = nativeReader.seService.version.toFloat()

    /**
     * Check if a card is present in this reader. see [Reader.isSecureElementPresent]
     * @return True if the card is present, false otherwise
     */
    override fun checkCardPresence(): Boolean {
        return nativeReader.isSecureElementPresent
    }

    /**
     * Get the card Answer To Reset
     * @return a byte array containing the ATR or null if no session was available
     */
    override fun getATR(): ByteArray? {
        return session?.let {
            Timber.i("Retrieving ATR from session...")
            it.atr
        }
    }

    /**
     * Open a logical channel by selecting the application
     * @param dfName A byte array containing the DF name or null if a basic opening is wanted.
     * @param isoControlMask The selection bits defined by the ISO selection command and expected by the OMAPI as P2 parameter.
     * @return A byte array containing the response to the OMAPI openLogicalChannel process or null if the Secure Element is unable to
     *         provide a new logical channel
     * @throws KeypleReaderIOException if the communication with the reader or the card has failed
     */
    @Throws(KeypleReaderIOException::class)
    override fun openChannelForAid(dfName: ByteArray?, isoControlMask: Byte): ByteArray? {
        if (dfName == null) {
            try {
                openChannel = session?.openBasicChannel(null)
            } catch (e: IOException) {
                Timber.e(e)
                throw KeypleReaderIOException("IOException while opening basic channel.")
            } catch (e: SecurityException) {
                Timber.e(e)
                throw KeypleReaderIOException("Error while opening basic channel, DFNAME = " + ByteArrayUtil.toHex(dfName), e.cause)
            }

            if (openChannel == null) {
                throw KeypleReaderIOException("Failed to open a basic channel.")
            }
        } else {
            Timber.i("[%s] openLogicalChannel => Select Application with AID = %s",
                    this.name, ByteArrayUtil.toHex(dfName))
            try {
                // openLogicalChannel of SimAlliance OMAPI is only available for version 3.0+ of the library.
                // By default the library always passes p2=00h
                // So if a p2 different of 00h is requested, we must check if omapi support it. Otherwise we throw an exception.
                val p2 = isoControlMask
                openChannel =
                        if (0 == p2.toInt()) {
                            session?.openLogicalChannel(dfName)
                        } else {
                            if (omapiVersion >= P2_SUPPORTED_MIN_VERSION) {
                                session?.openLogicalChannel(dfName, p2)
                            } else {
                                throw KeypleReaderIOException("P2 != 00h while opening logical channel is only supported by OMAPI version >= 3.0. Current is $omapiVersion")
                            }
                        }
            } catch (e: IOException) {
                Timber.e(e, "IOException")
                throw KeypleReaderIOException("IOException while opening logical channel.", e)
            } catch (e: NoSuchElementException) {
                Timber.e(e, "NoSuchElementException")
                throw java.lang.IllegalArgumentException(
                        "NoSuchElementException: " + ByteArrayUtil.toHex(dfName), e)
            } catch (e: SecurityException) {
                Timber.e(e, "SecurityException")
                throw KeypleReaderIOException("SecurityException while opening logical channel, aid :" + ByteArrayUtil.toHex(dfName), e.cause)
            }

            if (openChannel == null) {
                throw KeypleReaderIOException("Failed to open a logical channel.")
            }
        }
        /* get the FCI and build an ApduResponse */
        return openChannel!!.selectResponse
    }

    override fun isPhysicalChannelOpen(): Boolean {
        return session?.isClosed == false
    }

    @Throws(KeypleReaderIOException::class)
    public override fun openPhysicalChannel() {
        try {
            session = nativeReader.openSession()
        } catch (e: IOException) {
            Timber.e(e, "IOException")
            throw KeypleReaderIOException("IOException while opening physical channel.", e)
        }
    }

    /**
     * Close session see org.simalliance.openmobileapi.Session#close()
     */
    override fun closePhysicalChannel() {
        openChannel?.let {
            it.session.close()
            openChannel = null
        }
    }

    /**
     * Activates the provided card protocol.
     *
     *
     *  * Ask the plugin to take this protocol into account if a card using this protocol is
     * identified during the selection phase.
     *  * Activates the detection of SEs using this protocol (if the plugin allows it).
     *
     *
     * @param readerProtocolName The protocol to activate (must be not null).
     * @throws KeypleReaderProtocolNotSupportedException if the protocol is not supported.
     */
    override fun activateReaderProtocol(readerProtocolName: String?) {
        // Do nothing
    }

    /**
     * Deactivates the provided card protocol.
     *
     *
     *  * Ask the plugin to ignore this protocol if a card using this protocol is identified during
     * the selection phase.
     *  * Inhibits the detection of SEs using this protocol (if the plugin allows it).
     *
     *
     * @param readerProtocolName The protocol to deactivate (must be not null).
     */
    override fun deactivateReaderProtocol(readerProtocolName: String?) {
        // Do nothing
    }

    /**
     * Tells if the provided protocol matches the current protocol.
     *
     * @param readerProtocolName A not empty String.
     * @return True or false
     * @throws KeypleReaderProtocolNotFoundException if it is not possible to determine the protocol.
     * @since 1.0
     */
    override fun isCurrentProtocol(readerProtocolName: String?): Boolean {
        return AndroidOmapiSupportedProtocols.ISO_7816_3.name == readerProtocolName
    }

    /** Closes the logical channel explicitly.  */
    override fun closeLogicalChannel() {
        session?.closeChannels()
    }

    /**
     * Transmit an APDU command (as per ISO/IEC 7816) to the card see org.simalliance.openmobileapi.Channel#transmit(byte[])
     *
     * @param apduIn byte buffer containing the ingoing data
     * @return apduOut response
     * @throws KeypleReaderIOException if the communication with the reader or the card has failed
     */
    @Throws(KeypleReaderIOException::class)
    override fun transmitApdu(apduIn: ByteArray): ByteArray {
        // Initialization
        Timber.d("Data Length to be sent to tag : %s", apduIn.size)
        Timber.d("Data in : %s", ByteArrayUtil.toHex(apduIn))
        var dataOut = byteArrayOf(0)
        try {
            openChannel.let {
                dataOut = it?.transmit(apduIn) ?: throw IOException("Channel is not open")
            }
        } catch (e: IOException) {
            throw KeypleReaderIOException("Error while transmitting APDU", e)
        }

        Timber.d("Data out : " + ByteArrayUtil.toHex(dataOut))
        return dataOut
    }
}
